cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/DynamicSizing.java

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * Dynamic {@link Sizing} implementation that uses table size, minimum number of replicas, maximum partition size,
+ * and available Spark cores to determine the effective number of executor cores to use during the spark job execution.
+ *
+ * <p>This class is typically used when the table size is relatively small (few GBs). When reading small datasets,
+ * this class will allocate a limited number of resources to read the table. This in turn helps reduce the cost of
+ * coordinating a large number of executor cores when the dataset does not justify using the entire spark cluster
+ * for reading.
+ */
+public class DynamicSizing implements Sizing
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(DynamicSizing.class);
+
+    private final ReplicationFactor replicationFactor;
+    private final int maxPartitionSize;
+    private final int availableCores;
+    private final String keyspace;
+    private final String table;
+    private final String dc;
+    private final TableSizeProvider tableSizeProvider;
+    private final ConsistencyLevel consistencyLevel;
+
+    /**
+     * Constructs a new Sizing object.
+     *
+     * @param tableSizeProvider the table size provider
+     * @param consistencyLevel  the consistency level for the read operation
+     * @param replicationFactor the replication factor for the keyspace
+     * @param keyspace          the Cassandra keyspace
+     * @param table             the Cassandra table
+     * @param datacenter        the Cassandra datacenter
+     * @param maxPartitionSize  the maximum partition size desired
+     * @param availableCores    the maximum number of cores available
+     */
+    public DynamicSizing(TableSizeProvider tableSizeProvider,
+                         ConsistencyLevel consistencyLevel,
+                         ReplicationFactor replicationFactor,
+                         String keyspace,
+                         String table,
+                         String datacenter,
+                         int maxPartitionSize,
+                         int availableCores)
+    {
+        this.tableSizeProvider = tableSizeProvider;
+        this.consistencyLevel = consistencyLevel;
+        this.replicationFactor = replicationFactor;
+        this.keyspace = keyspace;
+        this.table = table;
+        this.dc = datacenter;
+        this.maxPartitionSize = maxPartitionSize;
+        this.availableCores = availableCores;
+    }
+
+    /**
+     * Returns the effective number of cores to be used during the spark execution.
+     * The value is calculated by getting the table size * the number of replicas
+     * we will use to read the data and then dividing it by the maximum partition
+     * size in GB. For example, assume we have a table with 7.25 GB of data, and
+     * assume a maximum partition size of 2.5 GB. Also, assume that a consistency
+     * level of 2. The number of cores is calculated by the following formula:
```

Comment:
I do not think it is "consistency level of 2", but should be 3. minReplicas is the quorum of 3.

Reply:
Correction: I meant to say, "the consistency level is quorum/local_quorum in this example, and RF is 3". It leads to the minReplicas of 2.